File: Prims.py
import heapq

def prim(graph, start):
    visited = set()
    min_heap = [(0, start)]
    total_cost = 0

    while min_heap:
        weight, node = heapq.heappop(min_heap)

        if node not in visited:
            visited.add(node)
            total_cost += weight
            print("Visited:", node, "Weight:", weight)

            for neighbor, w in graph[node]:
                if neighbor not in visited:
                    heapq.heappush(min_heap, (w, neighbor))

    print("Total Cost of MST:", total_cost)


# 🔹 Input Section
n = int(input("Enter number of vertices: "))
graph = {}

for i in range(n):
    node = input(f"Enter vertex {i+1}: ")
    edges = int(input(f"Enter number of neighbors for {node}: "))

    neighbors = []
    for j in range(edges):
        neighbor = input("Enter neighbor vertex: ")
        weight = int(input("Enter weight: "))
        neighbors.append((neighbor, weight))

    graph[node] = neighbors


start = input("Enter starting node: ")

# 🔹 Run Prim's Algorithm
prim(graph, start)

File: ai_practicals_cpp/prims.cpp
#include <iostream>

using namespace std;

int main()
{
    int n;

    cout << "Enter number of vertices: ";
    cin >> n;

    int graph[10][10];

    cout << "Enter adjacency matrix:\n";

    for(int i=0; i<n; i++)
    {
        for(int j=0; j<n; j++)
        {
            cin >> graph[i][j];
        }
    }

    int selected[10] = {0};

    selected[0] = 1;

    int edges = 0;

    cout << "\nEdges in MST:\n";

    while(edges < n-1)
    {
        int min = 9999;
        int x = 0;
        int y = 0;

        for(int i=0; i<n; i++)
        {
            if(selected[i])
            {
                for(int j=0; j<n; j++)
                {
                    if(!selected[j] && graph[i][j])
                    {
                        if(graph[i][j] < min)
                        {
                            min = graph[i][j];
                            x = i;
                            y = j;
                        }
                    }
                }
            }
        }

        cout << x << " - " << y
             << " : " << graph[x][y] << endl;

        selected[y] = 1;

        edges++;
    }

    return 0;
}
